Topic flag proton neutron number

n01r opened the pull request.

Ben-Schneider committed:
Add Server Sources

Provided by B. Schneider via R. Pausch
Tmp files removed and white spaces unified by A. Huebl

psychocoderHPC committed:
add message commands

- add FPS for simulation fps as command
- add RenderFPS for render fps as command

psychocoderHPC committed:
replace hard coded include and lib paths

ax3l committed:
Make Clean: No Errors

ax3l committed:
Remove PNGWriter (Feature was not implemented yet)

ax3l committed:
Correct VisLib Dependencies

ax3l committed:
Add new numGPUs/Cells/Particles IDs

ax3l committed:
Add GPL/LGPL Header

psychocoderHPC committed:
fix bug device selection

set device explicitly independent of the compute mode

ax3l committed:
Remove Dead Code

Commit:
Add flags + traits for proton + neutron number

Added:
 - particle flags (const attributes) for proton and neutron numbers
 - particle traits to get the proton and neutron numbers

Modified:
 - BSI calculation so that proton number is the maximum charge state

Commit:
Add \todo: Make proton and neutron number float

n01r commented:
Hey @ax3l and @psychocoderHPC 
Here's a pull request onto my main ionization branch which introduces the flags `protonNumber` and `neutronNumber`. Just tell me what you think. I can make changes and will start the real pull requests when #595 is merged.

PrometheusPi committed:
fix CUDA6.5 int(bool) bug in algorithms

ax3l commented on examples/LaserWakefield/cmakeFlags:
```diff
@@ -41,6 +41,7 @@ flags[8]="-DCUDA_ARCH=sm_20 -DPARAM_OVERWRITES:LIST=-DPARAM_DIMENSION=DIM2"
 flags[9]="-DCUDA_ARCH=sm_20 -DPARAM_OVERWRITES:LIST=-DPARAM_CURRENTSOLVER=currentSolver::Esirkepov<UsedParticleShape> ;-DPARAM_PARTICLESHAPE=particles::shapes::CIC;-DPARAM_DIMENSION=DIM2"
 flags[10]="-DCUDA_ARCH=sm_20 -DPARAM_OVERWRITES:LIST=-DPARAM_PRECISION=precision64Bit;-DPARAM_DIMENSION=DIM2"
 flags[11]="-DCUDA_ARCH=sm_35 -DPARAM_OVERWRITES:LIST=-DPARAM_GASPROFILE=gasNone;-DENABLE_CURRENT=0;-DPARAM_DIMENSION=DIM2"
+flags[12]="-DCUDA_ARCH=sm_20 -DPARAM_OVERWRITES:LIST=-DPARAM_GASPROFILE=gasSphereFlanks;-DENABLE_IONS=1;-DENABLE_IONIZATION=1"
```
call it `PARAM_IONS` ~~but keep `ENABLE_IONIZATION`~~ please.

I think `ENABLE_IONIZATION` is superfluous and can be removed.

ax3l commented on examples/LaserWakefield/include/simulation_defines/param/species.param:
```diff
@@ -34,7 +34,9 @@
 /*enable (1) or disable (0) electrons*/
 #define ENABLE_ELECTRONS 1
 /*enable (1) or disable (0) ions*/
+#ifndef ENABLE_IONS
```
``` c++
#ifndef PARAM_IONS
#define PARAM_IONS 0
#endif

#define ENABLE_IONS PARAM_IONS
```

n01r replied:
Even _this_ is still a nostalgic remnant. Actually from the moment on when we will have three or more species in the simulation this `#define` will become rather unsuitable.

ax3l replied:
absolutely

n01r replied:
So I guess I won't change this now because it is used in so many places (not by me) and it wouldn't be a very useful change if it is going to be changed later, anyway. 

If you insist, of course, I can replace all occurences of `ENABLE_IONS` and `ENABLE_ELECTRONS`.

ax3l replied:
no, just add the code above in exactly that one example (as explained offline)

ax3l commented on examples/LaserWakefield/include/simulation_defines/param/speciesDefinition.param:
```diff
@@ -0,0 +1,165 @@
+    shape<UsedParticleShape>,
+    interpolation<UsedField2Particle>,
+    current<UsedParticleCurrentSolver>
+    #if(ENABLE_IONIZATION==1)
```
I would always add them, no need to pragma around here.

if an ionization algorithm is called and they are not available, we should simply throw a static_assert.

n01r replied:
how? where? :sweat_smile:

n01r replied:
okay, found it ... let's see if what I build is nice

ax3l commented on src/picongpu/include/particles/traits/GetNeutronNumber.hpp:
```diff
@@ -0,0 +1,40 @@
+};
+} //namespace traits
+
+}// namespace picongpu
```
missing eof newline

ax3l commented on src/picongpu/include/particles/traits/GetProtonNumber.hpp:
```diff
@@ -0,0 +1,40 @@
+};
+} //namespace traits
+
+}// namespace picongpu
```
missing eof newline

ax3l commented on src/picongpu/include/particles/traits/GetNeutronNumber.hpp:
```diff
@@ -0,0 +1,40 @@
+template<typename T_Species>
+struct GetNeutronNumber
+{
+    typedef typename PMacc::traits::Resolve<
```
add a static assert here if the `T_Species` has no neutron number

ax3l commented on src/picongpu/include/particles/traits/GetProtonNumber.hpp:
```diff
@@ -0,0 +1,40 @@
+template<typename T_Species>
+struct GetProtonNumber
+{
+    typedef typename PMacc::traits::Resolve<
```
add a static assert here if the `T_Species` has no proton number

ax3l commented on src/picongpu/include/simulation_defines/param/speciesAttributes.param:
```diff
@@ -59,7 +59,8 @@ value_identifier(bool, radiationFlag, false);
  *
- * - only reasonable for ion species if ionization is enabled*/
+ * - only reasonable for ion species if ionization is enabled 
+ * - \todo maybe calculate from new attribute boundElectrons */
```
~~maybe~~

ax3l commented:
great!
just some minors how to add the attributes and how to react if the attribute is missing in a species' frame.

Commit:
Add static_assert if no proton/neutron number

If `GetProtonNumber` or `GetNeutronNumber` is called on a species that
does not have these flags, a static_assert error message is thrown.

Commit:
Remove pre-compiler defines for ionization

Ionization will be active at all times but the user can choose
if he/she wants to set `None` as the ionization model or simply
leave out the `ionizer` flag.

Commit:
Change cmakeFlags to enable ions in example

ax3l committed:
Particle2Grid: Non-Relativistic Particle Energy

PrometheusPi committed:
add missing TBG_waitJob variable

ax3l committed:
Add README.md with Install Notes

ax3l committed:
Merge pull request #670 from PrometheusPi/fix-k20AutoWait

fix k20_autoWait_profile.tpl

ax3l committed:
Fix Typo

PrometheusPi committed:
add analyser for numerical heating

This commit adds a python program that can compare the energy
evolution of two PIConGPU runs as suggested by @ax3l.

ax3l committed:
TBG_macros: Add More ADIOS Flags

Adds documentation about compression, aggregators and ost

ax3l committed:
Fix Typo in ADIOS Plugin Help

PrometheusPi committed:
Merge pull request #669 from ax3l/fix-lowEneDumpFld

Fix Particle2Grid: Non-Relativistic Particle Energy
looks good :+1:

psychocoderHPC committed:
Merge pull request #4 from ax3l/fix-deviceSelectionBug

Fix Typo

ax3l committed:
Merge pull request #595 from n01r/Multiple_Ionizers

Ionization in PIConGPU

ax3l committed:
Merge pull request #643 from psychocoderHPC/fix-deviceSelectionBug

fix device selection BUG

psychocoderHPC committed:
Merge pull request #673 from ax3l/doc-addMoreADIOSFlags

TBG_macros: Add More ADIOS Flags

ax3l committed:
Developers: Add Marco Garten

Welcome Marco!

PrometheusPi committed:
Merge pull request #676 from ax3l/doc-addMarco

Developers: Add Marco Garten

PrometheusPi committed:
Merge pull request #641 from ax3l/topic-addLiveVisServer

LiveVis: Add Server
I could compile and run it on hypnos.

psychocoderHPC committed:
fix adios zero amount writes

If some of the processes write zero amount of data with
adios they hang.

Changes:

- take care that all processes call collective adios write methods
- all processes calculate it's adios index table
- set freed memory to NULL
- move domain calculation before pre memory calculation
- call `adios_allocate_buffer()` every notify

PrometheusPi committed:
fix CUDA6.5 int(bool) bug in particle kernel

PrometheusPi committed:
add difference plot

PrometheusPi committed:
add absolute (not relative) plot option

PrometheusPi committed:
adjust sign convention when removing start value

This was suggested by @psychocoderHPC.
Additionally an unneeded legend was removed.

ax3l committed:
Merge pull request #661 from psychocoderHPC/fix-adiosZeroWrites

Fix ADIOS: Zero Writes, Collects and Buffers

psychocoderHPC committed:
fix MessageHeader, wrong array deletion

use `__deleteArray` to free memory

ax3l committed:
Merge pull request #672 from PrometheusPi/tool-numericalHeating

add analyzer tool for numerical heating

ax3l committed:
Merge pull request #683 from psychocoderHPC/fix-wrongDeleteInMessageHeader

fix MessageHeader, wrong array deletion

PrometheusPi committed:
adjust if condition as suggested by @psychocoderHPC

psychocoderHPC committed:
Merge pull request #657 from PrometheusPi/fix-02boolBugCUDA6.5

fix CUDA6.5 int(bool) bug in algorithms

psychocoderHPC committed:
Merge pull request #678 from PrometheusPi/fix-03boolBugCUDA6.5

fix CUDA6.5 int(bool) bug in particle kernel

Commit:
Make proton/neutronNumber a float

See:

https://github.com/ComputationalRadiationPhysics/picongpu/issues/650
